docs/source/schema-directives.md

```diff
@@ -1,179 +1,527 @@
+
+If you're using Apollo Server, you are also likely to be using the [`graphql-tools`](https://github.com/apollographql/graphql-tools) npm package, which provides a convenient yet powerful tool for implementing `@directive` syntax: the [`SchemaDirectiveVisitor`](https://github.com/apollographql/graphql-tools/blob/wip-schema-directives/src/schemaVisitor.ts) class.
+
+To implement a schema `@directive` using `SchemaDirectiveVisitor`, simply create a subclass of `SchemaDirectiveVisitor` that overrides one or more of the following visitor methods:
+
+* `visitSchema(schema: GraphQLSchema)`
+* `visitScalar(scalar: GraphQLScalarType)`
+* `visitObject(object: GraphQLObjectType)`
+* `visitFieldDefinition(field: GraphQLField<any, any>)`
+* `visitArgumentDefinition(argument: GraphQLArgument)`
+* `visitInterface(iface: GraphQLInterfaceType)`
+* `visitUnion(union: GraphQLUnionType)`
+* `visitEnum(type: GraphQLEnumType)`
+* `visitEnumValue(value: GraphQLEnumValue)`
+* `visitInputObject(object: GraphQLInputObjectType)`
+* `visitInputFieldDefinition(field: GraphQLInputField)`
+
+By overriding methods like `visitObject`, a subclass of `SchemaDirectiveVisitor` expresses interest in certain schema types such as `GraphQLObjectType` (the first parameter type of `visitObject`).
+
+When `SchemaDirectiveVisitor.visitSchemaDirectives` is called with a `GraphQLSchema` object and a map of visitor subclasses (`{ [directiveName: string]: typeof SchemaDirectiveVisitor }`), visitor methods overridden by those subclasses will be called with references to any schema type objects that have appropriately named `@directive`s attached to them, enabling the visitors to inspect or modify the schema.
+
+For example, if a directive called `@rest(url: "...")` appears after a field definition, a `SchemaDirectiveVisitor` subclass could provide meaning to that directive by overriding the `visitFieldDefinition` method (which receives a `GraphQLField` parameter), and then the body of that visitor method could manipulate the field's resolver function to fetch data from a REST endpoint:
+
+```typescript
+import {
+  makeExecutableSchema,
+  SchemaDirectiveVisitor,
+} from "graphql-tools";
 
-// Construct a schema, using GraphQL schema language
 const typeDefs = `
-  directive @upper on FIELD_DEFINITION
-
-  type Query {
-    hello: String @upper
-  }
-`;
-
-// Implement resolvers for out custom Directive
-const directiveResolvers = {
-  upper(
-    next,
-    src,
-    args,
-    context,
-  ) {
-    return next().then((str) => {
-      if (typeof(str) === 'string') {
-        return str.toUpperCase();
-      }
-      return str;
-    });
-  },
-}
+type Query {
+  people: [Person] @rest(url: "/api/v1/people")
+}`;
+
+const schema = makeExecutableSchema({ typeDefs });
+
+SchemaDirectiveVisitor.visitSchemaDirectives(schema, {
+  rest: class extends SchemaDirectiveVisitor {
```

Comment:
I think it would be better if these examples clearly separated the implementation, like so:

```js
class RestDirective extends SchemaDirectiveVisitor {
   ...
}

SchemaDirectiveVisitor.visitSchemaDirectives(schema, { rest: RestDirective });
```

That way it's clear that these are reusable across schemas, and I bet most people would want to define them in separate files and import them into where they use them.

Reply:
I ended up doing that in the examples down below, but I agree it would help here too.